clients/python-client/python_client/utils/kuberay_cluster_builder.py

```diff
@@ -0,0 +1,314 @@
+"""
+Set of helper methods to manage rayclusters. Requires Python 3.9 and higher
+"""
+
+import constants
+import copy
+import logging
+import math
+from typing import Any
+from abc import ABCMeta, abstractmethod
+from utils import kuberay_cluster_utils
+
+
+log = logging.getLogger(__name__)
+if logging.getLevelName(log.level) == "NOTSET":
+    logging.basicConfig(format="%(asctime)s %(message)s", level=constants.LOGLEVEL)
+
+
+class IClusterBuilder(metaclass=ABCMeta):
+    """
+    IClusterBuilder is an interface for building a cluster.
+
+    The class defines abstract methods for building the metadata, head pod, worker groups, and retrieving the built cluster.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def build_meta():
+        "builds the cluster metadata"
+
+    @staticmethod
+    @abstractmethod
+    def build_head():
+        "builds the head pod"
+
+    @staticmethod
+    @abstractmethod
+    def build_worker():
+        "builds a worker group"
+
+    @staticmethod
+    @abstractmethod
+    def get_cluster():
+        "Returns the built cluster"
+
+
+# Concrete implementation of the builder interface
+class ClusterBuilder(IClusterBuilder):
+    """
+    ClusterBuilder implements the abstract methods of IClusterBuilder to build a cluster.
+    """
+
+    def __init__(self):
+        self.cluster: dict[str, Any] = {}
+        self.succeeded: bool = False
+        self.cluster_utils = kuberay_cluster_utils.ClusterUtils()
+
+    def build_meta(
+        self,
+        name: str,
+        k8s_namespace: str = "default",
+        labels: dict = None,
+        ray_version: str = "2.2.0",
+    ):
+        """Builds the metadata and ray version of the cluster.
+
+        Parameters:
+        - name (str): The name of the cluster.
+        - k8s_namespace (str, optional): The namespace in which the Ray cluster exists. Defaults to "default".
+        - labels (dict, optional): A dictionary of key-value pairs to add as labels to the cluster. Defaults to None.
+        - ray_version (str, optional): The version of Ray to use for the cluster. Defaults to "2.2.0".
+        """
+        self.cluster = self.cluster_utils.populate_meta(
+            cluster=self.cluster,
+            name=name,
+            k8s_namespace=k8s_namespace,
+            labels=labels,
+            ray_version=ray_version,
+        )
+        return self
+
+    def build_head(
+        self,
+        ray_image: str = "rayproject/ray:2.2.0",
+        service_type: str = "ClusterIP",
+        cpu_requests: str = "1",
+        memory_requests: str = "1G",
```

Comment:
We don't recommend less than 2Gi memory for the Ray head these days.
Maybe we can match the requests and limits here?

I don't think there's any universal sense of good default resource values for a Ray node.

Reply:
fixed now with 3G